### PR TITLE
Merge cherry-picked s3 ufs changes from 2.x branch

### PR DESCRIPTION
**Pull request by elega**
### What changes are proposed in this pull request?

Partially cherry picked changes from 
https://github.com/Alluxio/alluxio/pull/17102 and
https://github.com/Alluxio/alluxio/pull/17242

1. Introduced s3 mock for unit tests & corresponding unit tests for s3
2. UFS ListStatus iterator interface
3.  Async S3 listing & get interfaces & integration with the new s3 client 

### Why are the changes needed?

Better performance & memory control for s3

### Does this PR introduce any user facing changes?

N/A

### CODE REVIEW

**Comment by elega** [comment]
alluxio-bot, check this please

File: dora/core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java [hunk 1]
```diff
@@ -163,6 +189,94 @@ public boolean isSeekable() {
...
+  public Iterator<UfsStatus> listStatusIterable(
+      String path, ListOptions options, String startAfter, int batchSize) throws IOException {
+    // Calling this method on non s3 UFS might result in OOM because batch based fetching
+    // is not supported and this method essentially fetches all ufs status and converts it to
+    // an iterator.
+    UfsStatus[] result = listStatus(path, options);
+    if (result == null) {
+      return null;
+    }
+    Arrays.sort(result, Comparator.comparing(UfsStatus::getName));
```
Comment (jja725) [inline dora/core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:203]: Is this sort required for non s3 ufs?

**Review by jja725 — APPROVED** [review]
listStatusIterable part LGTM, leave some questions & comments but those don't block merge

File: dora/core/common/src/main/java/alluxio/underfs/UnderFileSystem.java [hunk 2]
```diff
@@ -59,7 +60,7 @@
 @PublicApi
 @ThreadSafe
 // TODO(adit); API calls should use a URI instead of a String wherever appropriate
-public interface UnderFileSystem extends Closeable {
+public interface UnderFileSystem extends Closeable, AsyncUfsClient {
```
[the remark this answers is not in the document]
Comment (elega) [reply dora/core/common/src/main/java/alluxio/underfs/UnderFileSystem.java:63]: metadata sync redesign, which will be picked up later, uses this interface for the ease of mocking

File: dora/core/common/src/main/java/alluxio/underfs/UnderFileSystemWithLogging.java [hunk 3]
```diff
@@ -819,6 +824,38 @@ public String toString() {
     });
   }
 
+  @Override
+  public Iterator<UfsStatus> listStatusIterable(
+      String path, ListOptions options, String startAfter,
+      int batchSize) throws IOException {
+    return call(new UfsCallable<Iterator<UfsStatus>>() {
```
[the remark this answers is not in the document]
Comment (elega) [reply dora/core/common/src/main/java/alluxio/underfs/UnderFileSystemWithLogging.java:831]: will change in the next PR

File: dora/core/common/src/main/java/alluxio/util/RateLimiter.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+import java.util.Optional;
+
+/**
+ * Used to limit the rate of operations. This rate limiter is not thread safe
+ * and the operations are non-blocking. It is used by acquiring a permit for
+ * each operation, then checking how the operation should wait by calling
+ * {@link RateLimiter#getWaitTimeNanos(long)}.
+ */
+public interface RateLimiter {
```
[the remark this answers is not in the document]
Comment (elega) [reply dora/core/common/src/main/java/alluxio/util/RateLimiter.java:22]: no Tyler introduced this and the behavior is different from the guava one. I once asked him the same question.

File: dora/core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java (hunk 1 above)
Comment (elega) [reply dora/core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:203]: not necessarily but we can just let the interface always return a sorted result. I can probably update the interface desc later.

Comment (elega) [reply dora/core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java:203]: actually let me remove this

Comment (elega) [reply dora/core/common/src/main/java/alluxio/underfs/BaseUnderFileSystem.java]: actually i removed this

**Comment by elega** [comment]
alluxio-bot, merge this please

**Pull request merged**